### PR TITLE
Implemented Idint

### PR DESCRIPTION
**Pull request by parth121101**
Implemented Idint

### CODE REVIEW

File: src/libasr/pass/intrinsic_function_registry.h [hunk 1]
```diff
@@ -2100,6 +2102,76 @@ namespace Sngl {
...
+            "ASR Verify: Arguments to `idint` must be of real type",
+            x.base.base.loc, diagnostics);
+    }
+
+    static ASR::expr_t *eval_Idint(Allocator &al, const Location &loc,
+            ASR::ttype_t* arg_type, Vec<ASR::expr_t*> &args) {
+        double v = ASR::down_cast<ASR::RealConstant_t>(expr_value(args[0]))->m_r;
+        // cast it down correctly to integer
+        int val = int(v);
+        return f(val, arg_type);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:2124]:
```suggestion
        int val = ASR::down_cast<ASR::RealConstant_t>(expr_value(args[0]))->m_r;
        return make_ConstantWithType(make_IntegerConstant_t, val, ASRUtils::TYPE(ASR::make_Integer_t(al, loc, 4)), loc);
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 2]
```diff
@@ -2100,6 +2102,76 @@ namespace Sngl {
...
+
+    static inline void verify_args(const ASR::IntrinsicScalarFunction_t& x,
+            diag::Diagnostics& diagnostics) {
+        ASRUtils::require_impl(x.n_args == 1,
+            "ASR Verify: Call `idint` must have exactly one argument",
+            x.base.base.loc, diagnostics);
+        ASR::ttype_t *type = ASRUtils::expr_type(x.m_args[0]);
+        int kind = ASRUtils::extract_kind_from_ttype_t(type);
+        ASRUtils::require_impl(ASRUtils::is_real(*type) && kind == 8,
+            "ASR Verify: Arguments to `idint` must be of real type",
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:2115]:
```suggestion
            "ASR Verify: Arguments to `idint` must be of double precision type",
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 3]
```diff
@@ -2100,6 +2102,76 @@ namespace Sngl {
...
+    }
+
+    static inline ASR::asr_t* create_Idint(
+            Allocator& al, const Location& loc, Vec<ASR::expr_t*>& args,
+            const std::function<void (const std::string &, const Location &)> err) {
+        ASR::ttype_t* return_type = int32;
+        if ( args.n != 1 ) {
+            err("Intrinsic `idint` accepts exactly one argument", loc);
+        } else if ( !is_real(*expr_type(args[0])) ) {
+            err("Argument of the `idint` must be Real", loc);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:2134]:
```suggestion
            err("Argument of the `idint` must be Double Precision", loc);
```

File: integration_tests/intrinsics_128.f90 [hunk 4]
```diff
@@ -0,0 +1,7 @@
+program intrinsics_128
+    double precision :: x
+    x = 4.23
+
+    print *, idint(x)
+    if( .not. idint(x) == 4 ) error stop
```
Comment (certik) [inline integration_tests/intrinsics_128.f90:6]: This is a runtime test, which we need. Can you please also add a compile time test (`idint(4.23)`)?

**Comment by certik** [comment]
Otherwise this looks good.

**Comment by Pranavchiku** [comment]
Closing with reference to https://github.com/lfortran/lfortran/pull/3193

**Pull request closed without merging**